src/server/worker/datum/worker.go

```diff
@@ -0,0 +1,385 @@
+		fsi := NewFileSetIterator(pachClient, task.FileSetId, task.PathRange)
+		keyHasher := pfs.NewHash()
+		marshaller := new(jsonpb.Marshaler)
+		err := fsi.Iterate(func(meta *Meta) error {
+			for _, input := range meta.Inputs {
+				var rawKey string
+				switch task.Type {
+				case KeyTask_JOIN:
+					rawKey = input.JoinOn
+				case KeyTask_GROUP:
+					rawKey = input.GroupBy
+				default:
+					return errors.Errorf("unrecognized key task type (%v)", task.Type)
+				}
+				// hash input keys to ensure consistently-shaped filepaths
+				keyHasher.Reset()
+				keyHasher.Write([]byte(rawKey))
+				key := hex.EncodeToString(keyHasher.Sum(nil))
+				out, err := marshaller.MarshalToString(input)
+				if err != nil {
+					return errors.Wrap(err, "marshalling input for key aggregation")
+				}
+				if err := mf.PutFile(key, strings.NewReader(out), client.WithAppendPutFile()); err != nil {
+					return errors.EnsureStack(err)
+				}
+			}
+			return nil
+		})
+		return errors.EnsureStack(err)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return serializeKeyTaskResult(&KeyTaskResult{FileSetId: resp.FileSetId})
+}
+
+func processMergeTask(pachClient *client.APIClient, task *MergeTask) (*types.Any, error) {
+	fileSetID, err := WithCreateFileSet(pachClient, "pachyderm-datums-merge", func(s *Set) error {
+		var fsmis []Iterator
+		for _, fileSetId := range task.FileSetIds {
+			fsmis = append(fsmis, newFileSetMultiIterator(pachClient, fileSetId, task.PathRange))
+		}
+		switch task.Type {
+		case MergeTask_JOIN:
+			return mergeByKey(fsmis, existingMetaHash, func(metas []*Meta) error {
+				var crossInputs [][]*common.Input
+				for _, m := range metas {
+					crossInputs = append(crossInputs, m.Inputs)
+				}
+				err := newCrossListIterator(crossInputs).Iterate(func(meta *Meta) error {
+					if len(meta.Inputs) == len(task.FileSetIds) {
+						// all inputs represented, include all inputs
+						return s.UploadMeta(meta)
+					}
+					var filtered []*common.Input
+					for _, in := range meta.Inputs {
+						if in.OuterJoin {
+							filtered = append(filtered, in)
+						}
+					}
+					if len(filtered) > 0 {
+						return s.UploadMeta(&Meta{Inputs: filtered})
+					}
+					return nil
+				})
+				return errors.EnsureStack(err)
+			})
+		case MergeTask_GROUP:
+			return mergeByKey(fsmis, existingMetaHash, func(metas []*Meta) error {
+				var allInputs []*common.Input
+				for _, m := range metas {
+					allInputs = append(allInputs, m.Inputs...)
+				}
+				return s.UploadMeta(&Meta{Inputs: allInputs})
+			})
+		default:
+			return errors.Errorf("unrecognized merge task type (%v)", task.Type)
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+	return serializeMergeTaskResult(&MergeTaskResult{FileSetId: fileSetID})
+}
+
+func existingMetaHash(meta *Meta) string {
+	return meta.Hash
+}
+
+func processComposeTask(pachClient *client.APIClient, task *ComposeTask) (*types.Any, error) {
```

Comment:
Yes, the reason I created a task for this is to ensure cache hits through the creation process. Cache hits are based on task ids, which are based on the values of the fields in the proto. If we didn't cache this step, then we would get a different file set id at this point, which would cause us to miss all of the subsequent cache entries. This is more generally an issue with file set ids not being stable which we have some plans for, but probably not the near future.

Reply:
ok that makes sense. WDYT about adding some of that in a comment?